File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.swift
import Foundation

func solution(_ dice:[[Int]]) -> [Int] {
    let n = dice.count
    var maxWin = -1
    var maxCombi: [Int] = []
    
    let allCombi = getCombinations(of: Array(0..<n), n/2)
    
    for combi in allCombi {
        
        let A = combi
        let B = Array(Set(0..<n).subtracting(A))
        
        let sumsA = getSum(of: A, dice: dice)
        let sumsB = getSum(of: B, dice: dice).sorted()
        
        var win = 0
        for sumA in sumsA {
            win += lowerBound(of: sumsB, target: sumA)
        }
        
        if win > maxWin {
            maxWin = win
            maxCombi = combi
        }
    }
    
    return maxCombi.map { $0 + 1 }
}

func lowerBound(of arr: [Int], target: Int) -> Int {
    var st = 0
    var en = arr.count
    
    while st < en {
        let mid = (st + en) / 2
        
        if arr[mid] < target { st = mid + 1 }
        else { en = mid }
    }
    
    return st
}

func getSum(of arr: [Int], dice: [[Int]]) -> [Int] {
    // arr: 선택한 주사위의 인덱스 배열
    var sums: [Int] = [] // 여기에 나올 수 있는 눈의 합의 모든 경우의 수
    
    var curDices: [[Int]] = []
    for idx in arr { curDices.append(dice[idx]) }
    
    var sum = 0
    func permu(_ n: Int) {
        if n == curDices.count {
            sums.append(sum)
            return
        }
        
        for i in 0..<6 {
            sum += curDices[n][i]
            permu(n+1)
            sum -= curDices[n][i]
        }
    }
    permu(0)
    
    return sums
}

func getCombinations<T>(of elements: [T], _ k: Int) -> [[T]] {
    var results: [[T]] = []
    
    var selected: [T] = []
    func comb(_ n: Int, _ start: Int) {
        if n == k {
            results.append(selected)
            return
        }
        
        for i in start..<elements.count {
            selected.append(elements[i])
            comb(n+1, i+1)
            selected.removeLast()
        }
    }
    comb(0,0)
    
    return results
}

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.py
# @time -1

from itertools import combinations

def cal_sum(list_dice):
    
    sums = []
    
    def dfs(k, s, sums):
        
        if k == len(list_dice):
            sums.append(s)
            return
        
        for i in range(6):
            dfs(k+1, s+list_dice[k][i], sums)
    
    dfs(0, 0, sums)
            
    return sums

def lower_bound(target, sums):
    
    st = 0
    en = len(sums)
    
    while st < en:
        mid = (st + en) // 2
        
        if sums[mid] < target:
            st = mid + 1
        else:
            en = mid
            
    return st

def solution(dice):
    
    n = len(dice)
    max_win_dices = []
    max_win_cnt = 0
    
    for dice_nos in map(list, combinations(range(n), n // 2)):
        
        set_a = set(dice_nos)
        set_b = set(range(n)) - set_a
        
        list_a = list(map(lambda x: dice[x], set_a))
        list_b = list(map(lambda x: dice[x], set_b))
            
        sum_a = sorted(cal_sum(list_a))
        sum_b = sorted(cal_sum(list_b))
        
        win_cnt = 0
        
        for s in sum_a:
            win_cnt += lower_bound(s, sum_b)
            
        if win_cnt > max_win_cnt:
            max_win_cnt = win_cnt
            max_win_dices = [i + 1 for i in dice_nos]
        
    return max_win_dices